usm: Allow using eventstream instead of netlink

vitkyrka opened the pull request:
<!--
* New contributors are highly encouraged to read our
  [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* Both Contributor and Reviewer Checklists are available at https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#pull-requests.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.
-->
### What does this PR do?

Adds an option (disabled by default) to make the process monitor used by USM to use the event stream instead of netlink to receive process events
.
Several unit tests have been modified to run with both netlink and event stream. 

<!--
* A brief description of the change being made with this pull request.
* If the description here cannot be expressed in a succinct form, consider
  opening multiple pull requests instead of a single one.
-->

### Motivation

https://datadoghq.atlassian.net/browse/USMON-813

<!--
* What inspired you to submit this pull request?
* Link any related GitHub issues or PRs here.
-->

### Additional Notes

https://datadoghq.atlassian.net/wiki/spaces/UT/pages/3668248553/

<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->

### Possible Drawbacks / Trade-offs

<!--
* What are the possible side-effects or negative impacts of the code change?
-->

### Describe how to test/QA your changes

Unit tests are added.

<!--
* Write here in detail or link to detailed instructions on how this change can
  be tested/QAd/validated, including any environment setup.
-->

vitkyrka committed:
usm: Allow using eventstream instead of netlink

vitkyrka committed:
usm: process monitor: Make test wait for exit events

The test asserts that exec and exit events were received but only waits
for the former, so there is a race condition which leads to the
assertion of the exit events failing.  Make the test wait for both
events.

vitkyrka committed:
event monitor: Skip test on unsupported kernels

vitkyrka committed:
usm: process monitor: Only build testutil on Linux

The package it self is only built on Linux so do not attempt to build
the testutil on other paltforms.

vitkyrka committed:
system-probe: eventmonitor: Only register USM consumer on Linux

USM's process monitor is only built on Linux so register that consumer
from a Linux-specific file, to avoid build failures on other platforms.

vitkyrka committed:
event monitor: Only build testutil on Linux/Windows

To match the package and avoid build errors on macOS.

vitkyrka committed:
Merge remote-tracking branch 'origin/main' into eventstream

guyarb commented on pkg/eventmonitor/testutil/testutil.go:
```diff
@@ -0,0 +1,46 @@
+	"os"
+	"testing"
+
+	sysconfig "github.com/DataDog/datadog-agent/cmd/system-probe/config"
+	"github.com/DataDog/datadog-agent/comp/core/workloadmeta"
+	"github.com/DataDog/datadog-agent/pkg/eventmonitor"
+	emconfig "github.com/DataDog/datadog-agent/pkg/eventmonitor/config"
+	secconfig "github.com/DataDog/datadog-agent/pkg/security/config"
+	"github.com/DataDog/datadog-agent/pkg/util/optional"
+	"github.com/stretchr/testify/require"
```
nit - split the imports to 3rd party imports, and datadog only

guyarb commented on pkg/eventmonitor/testutil/testutil.go:
```diff
@@ -0,0 +1,46 @@
+	require.NoError(t, err)
+
+	// Needed for the socket creation to work
+	require.NoError(t, os.MkdirAll("/opt/datadog-agent/run/", 0755))
```
how's that supported on windows?

hmahmood commented on cmd/system-probe/modules/eventmonitor_linux.go:
```diff
@@ -10,15 +10,40 @@ package modules
 	},
 }
+
+func createEventMonitorModuleLinux(config *sysconfigtypes.Config, wmeta optional.Option[workloadmeta.Component]) (module.Module, error) {
```
Is there a need for the `Linux` prefix? We're already in a file that is only going to get compiled on linux.

hmahmood replied:
I think it might be cleaner if you had a function that created the procmon monitor only (on windows that would just return nil) and then called it from `createEventMonitor`.

hmahmood commented on pkg/process/monitor/process_monitor.go:
```diff
@@ -347,46 +349,56 @@ func (pm *ProcessMonitor) mainEventLoop() {
-			pm.netlinkEventsChannel = make(chan netlink.ProcEvent, processMonitorEventQueueSize)
 
-			go pm.mainEventLoop()
+			if !useEventStream {
```
```go
if useEventStream {
  return
}
```

hmahmood commented on pkg/process/monitor/process_monitor.go:
```diff
@@ -485,3 +497,83 @@ func FindDeletedProcesses[V any](pids map[uint32]V) map[uint32]struct{} {
+// EventConsumer defines an event consumer to handle event monitor events in the
+// process monitor
+type EventConsumer struct {
+	sync.RWMutex
```
What is the lock for?

vitkyrka committed:
system-probe: eventmonitor: Refactor platform-specific init

Create a per-platform function called from the common function instead
of the other way around.

vitkyrka committed:
event monitor: Only build testutil on Linux

The code is only used on Linux.

vitkyrka committed:
event monitor: testutil: Split third-party imports

vitkyrka committed:
process monitor: Refactor condition

Exit early to reduce nesting level.

vitkyrka committed:
process monitor: Rename receiver

vitkyrka committed:
process monitor: Remove extra variable

vitkyrka committed:
process monitor: testutil: Separate third-party imports

vitkyrka committed:
process monitor: testutil: Add test build flag

vitkyrka committed:
process monitor: Remove lock

vitkyrka committed:
Merge remote-tracking branch 'origin/main' into eventstream

vitkyrka committed:
event monitor: Rename unused parameter

guyarb commented on pkg/process/monitor/process_monitor.go:
```diff
@@ -347,16 +349,28 @@ func (pm *ProcessMonitor) mainEventLoop() {
 		func() {
-			log.Info("initializing process monitor")
+			var method string
+			if useEventStream {
+				method = "event stream"
+			} else {
+				method = "netlink"
+			}
+			log.Infof("initializing process monitor (%s)", method)
 			pm.tel = newProcessMonitorTelemetry()
-			pm.done = make(chan struct{})
+
 			pm.initCallbackRunner()
 
 			pm.processMonitorWG.Add(1)
+
+			if useEventStream {
+				return
+			}
+
+			pm.done = make(chan struct{})
```
I'm afraid the changes here are going to create deadlocks and issues during shutdown (not deterministically, but during races)

hmahmood replied:
`Stop` seems to be overloaded with different sync primitives. Any chance we can clean that up?

hmahmood replied:
Would moving this line back to where it was solve the issue for the time being?

hmahmood commented on pkg/process/monitor/process_monitor.go:
```diff
@@ -347,16 +349,28 @@ func (pm *ProcessMonitor) mainEventLoop() {
 	pm.initOnce.Do(
 		func() {
-			log.Info("initializing process monitor")
+			var method string
```
nit: 
```go
method := "netlink"
if useEventStream {
  method = "event stream"
}
```

vitkyrka committed:
process monitor: Move WaitGroup init

The processMonitorWG is not used in the event stream case, make
this clearer by only doing the Add on it for netlink.

vitkyrka committed:
process monitor: Stop callback runners for event stream too

Make sure the callback runners are stopped when Stop() is
called even for the event stream case.  Otherwise there is
a race (detected by the go race detector) when the
TestProcessMonitor test suite is run.

vitkyrka committed:
Merge remote-tracking branch 'origin/main' into eventstream

vitkyrka committed:
process monitor: Simplify variable init

vitkyrka commented:
/trigger-ci --variable RUN_UNIT_TESTS=on

brycekahle reviewed (approved):
👍 for ebpf-platform owned files

vitkyrka commented:
/merge
